File: sdks/go/rodo.go
package rodo

import (
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"os"
	"time"
)

const (
	magic       uint32 = 0x52443031
	versionMajor byte = 1
	versionMinor byte = 0
)

// Value types
const (
	TypeNull   byte = 0x00
	TypeBool   byte = 0x01
	TypeInt    byte = 0x02
	TypeFloat  byte = 0x03
	TypeString byte = 0x04
	TypeBytes  byte = 0x05
	TypeArray  byte = 0x06
	TypeMap    byte = 0x07
	TypeDate   byte = 0x08
	TypeUUID   byte = 0x09
)

// Value represents a Rodo value.
type Value struct {
	Type    byte
	BoolVal bool
	IntVal  int64
	FloatVal float64
	StringVal string
	BytesVal []byte
	ArrayVal []Value
	MapVal   map[string]Value
	DateVal  int64
	UUIDVal  [16]byte
}

// Rodo represents a connection to a .rd file.
type Rodo struct {
	file       *os.File
	symbols    map[string]uint32
	symbolNext uint32
	dict       map[uint32]string
	dictNext   uint32
	pairs      map[uint32]Value // single block, block ID = 1
	dirty      bool
}

// Link opens or creates a .rd file at path.
func Link(path string) (*Rodo, error) {
	f, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0644)
	if err != nil {
		return nil, err
	}
	r := &Rodo{
		file:       f,
		symbols:    make(map[string]uint32),
		symbolNext: 0,
		dict:       make(map[uint32]string),
		dictNext:   0,
		pairs:      make(map[uint32]Value),
		dirty:      false,
	}
	// Check if file is non-empty
	stat, err := f.Stat()
	if err != nil {
		return nil, err
	}
	if stat.Size() > 0 {
		if err := r.load(); err != nil {
			f.Close()
			return nil, err
		}
	}
	return r, nil
}

// Close writes changes and closes the file.
func (r *Rodo) Close() error {
	if r == nil {
		return nil
	}
	if r.dirty {
		if err := r.save(); err != nil {
			r.file.Close()
			return err
		}
	}
	return r.file.Close()
}

// Set stores a value for a key.
func (r *Rodo) Set(key string, val Value) error {
	symID, ok := r.symbols[key]
	if !ok {
		r.symbolNext++
		symID = r.symbolNext
		r.symbols[key] = symID
	}
	// If value is string, add to dictionary
	if val.Type == TypeString {
		if _, exists := r.dictFind(val.StringVal); !exists {
			r.dictNext++
			r.dict[r.dictNext] = val.StringVal
		}
	}
	// If map, need to add keys to dict (simplified: ignore)
	r.pairs[symID] = val
	r.dirty = true
	return nil
}

// Get returns the value for a key, or nil if not found.
func (r *Rodo) Get(key string) (Value, bool) {
	symID, ok := r.symbols[key]
	if !ok {
		return Value{}, false
	}
	v, exists := r.pairs[symID]
	return v, exists
}

// Has returns true if the key exists.
func (r *Rodo) Has(key string) bool {
	symID, ok := r.symbols[key]
	if !ok {
		return false
	}
	_, exists := r.pairs[symID]
	return exists
}

// Delete removes a key.
func (r *Rodo) Delete(key string) error {
	symID, ok := r.symbols[key]
	if !ok {
		return nil
	}
	if _, exists := r.pairs[symID]; exists {
		delete(r.pairs, symID)
		r.dirty = true
	}
	return nil
}

// Keys returns all keys.
func (r *Rodo) Keys() []string {
	var keys []string
	for name, id := range r.symbols {
		if _, exists := r.pairs[id]; exists {
			keys = append(keys, name)
		}
	}
	return keys
}

// All returns a map of all keys to values.
func (r *Rodo) All() map[string]Value {
	result := make(map[string]Value)
	for name, id := range r.symbols {
		if v, exists := r.pairs[id]; exists {
			result[name] = v
		}
	}
	return result
}

// ---------------- Internal functions ----------------

func (r *Rodo) dictFind(s string) (uint32, bool) {
	for id, str := range r.dict {
		if str == s {
			return id, true
		}
	}
	return 0, false
}

func (r *Rodo) load() error {
	// Seek to beginning
	if _, err := r.file.Seek(0, io.SeekStart); err != nil {
		return err
	}
	data, err := io.ReadAll(r.file)
	if err != nil {
		return err
	}
	if len(data) < 18 {
		return errors.New("arquivo .rd inválido")
	}
	if binary.BigEndian.Uint32(data[0:4]) != magic {
		return errors.New("magic number inválido")
	}
	offset := 18 // skip header

	// Read symbols
	symbolCount, err := readVarint(data, &offset)
	if err != nil {
		return err
	}
	for i := uint64(0); i < symbolCount; i++ {
		id, err := readVarint(data, &offset)
		if err != nil {
			return err
		}
		nameLen, err := readVarint(data, &offset)
		if err != nil {
			return err
		}
		name := string(data[offset : offset+int(nameLen)])
		offset += int(nameLen)
		// skip typeHint and flags
		offset += 2
		r.symbols[name] = uint32(id)
		if uint32(id) > r.symbolNext {
			r.symbolNext = uint32(id)
		}
	}

	// Read dictionary
	dictCount, err := readVarint(data, &offset)
	if err != nil {
		return err
	}
	for i := uint64(0); i < dictCount; i++ {
		id, err := readVarint(data, &offset)
		if err != nil {
			return err
		}
		strLen, err := readVarint(data, &offset)
		if err != nil {
			return err
		}
		str := string(data[offset : offset+int(strLen)])
		offset += int(strLen)
		r.dict[uint32(id)] = str
		if uint32(id) > r.dictNext {
			r.dictNext = uint32(id)
		}
	}

	// Read blocks (only one)
	blockCount, err := readVarint(data, &offset)
	if err != nil {
		return err
	}
	if blockCount > 0 {
		_, err := readVarint(data, &offset) // block ID
		if err != nil {
			return err
		}
		pairCount, err := readVarint(data, &offset)
		if err != nil {
			return err
		}
		for i := uint64(0); i < pairCount; i++ {
			symID, err := readVarint(data, &offset)
			if err != nil {
				return err
			}
			val, err := readValue(data, &offset, r.dict)
			if err != nil {
				return err
			}
			r.pairs[uint32(symID)] = val
		}
	}
	return nil
}

func (r *Rodo) save() error {
	var buf bytes.Buffer

	// Header
	buf.Write([]byte{byte(magic >> 24), byte(magic >> 16), byte(magic >> 8), byte(magic)})
	buf.WriteByte(versionMajor)
	buf.WriteByte(versionMinor)
	buf.WriteByte(0) // flags
	// Timestamp (8 bytes) - use current Unix ms
	var tsBytes [8]byte
	binary.BigEndian.PutUint64(tsBytes[:], uint64(time.Now().UnixNano()/1e6))
	buf.Write(tsBytes[:])
	// CRC32 placeholder
	buf.Write([]byte{0, 0, 0, 0})

	// Symbols
	writeVarint(&buf, uint64(len(r.symbols)))
	for name, id := range r.symbols {
		writeVarint(&buf, uint64(id))
		writeVarint(&buf, uint64(len(name)))
		buf.WriteString(name)
		buf.WriteByte(0) // typeHint
		buf.WriteByte(0) // flags
	}

	// Dictionary: collect strings from values
	dictMap := make(map[string]uint32)
	for _, val := range r.pairs {
		collectStrings(val, &dictMap, &r.dictNext)
	}
	// Override r.dict with collected
	r.dict = make(map[uint32]string)
	for s, id := range dictMap {
		r.dict[id] = s
	}
	writeVarint(&buf, uint64(len(r.dict)))
	// Write dict sorted by ID
	ids := make([]uint32, 0, len(r.dict))
	for id := range r.dict {
		ids = append(ids, id)
	}
	sort.Slice(ids, func(i, j int) bool { return ids[i] < ids[j] })
	for _, id := range ids {
		writeVarint(&buf, uint64(id))
		s := r.dict[id]
		writeVarint(&buf, uint64(len(s)))
		buf.WriteString(s)
	}

	// Blocks
	writeVarint(&buf, 1) // block count
	writeVarint(&buf, 1) // block ID
	writeVarint(&buf, uint64(len(r.pairs)))
	for symID, val := range r.pairs {
		writeVarint(&buf, uint64(symID))
		if err := writeValue(&buf, val, r.dict); err != nil {
			return err
		}
	}

	// Metadata
	writeVarint(&buf, uint64(len(r.pairs)))
	writeVarint(&buf, uint64(len(r.symbols)))
	writeVarint(&buf, uint64(len(r.dict)))
	writeVarint(&buf, uint64(buf.Len()))
	buf.Write([]byte{0, 0, 0, 0}) // CRC32 placeholder

	// Write to file
	if _, err := r.file.Seek(0, io.SeekStart); err != nil {
		return err
	}
	if err := r.file.Truncate(0); err != nil {
		return err
	}
	if _, err := r.file.Write(buf.Bytes()); err != nil {
		return err
	}
	r.dirty = false
	return nil
}

func collectStrings(val Value, dict *map[string]uint32, nextID *uint32) {
	if val.Type == TypeString {
		if _, exists := (*dict)[val.StringVal]; !exists {
			*nextID++
			(*dict)[val.StringVal] = *nextID
		}
	} else if val.Type == TypeMap {
		for k, v := range val.MapVal {
			if _, exists := (*dict)[k]; !exists {
				*nextID++
				(*dict)[k] = *nextID
			}
			collectStrings(v, dict, nextID)
		}
	} else if val.Type == TypeArray {
		for _, item := range val.ArrayVal {
			collectStrings(item, dict, nextID)
		}
	}
}

func writeVarint(buf *bytes.Buffer, value uint64) {
	for value >= 0x80 {
		buf.WriteByte(byte(value) | 0x80)
		value >>= 7
	}
	buf.WriteByte(byte(value))
}

func readVarint(data []byte, offset *int) (uint64, error) {
	var result uint64
	var shift uint
	for {
		if *offset >= len(data) {
			return 0, io.ErrUnexpectedEOF
		}
		b := data[*offset]
		*offset++
		result |= uint64(b&0x7F) << shift
		shift += 7
		if b&0x80 == 0 {
			break
		}
		if shift > 70 {
			return 0, errors.New("varint muito longo")
		}
	}
	return result, nil
}

func zigzagEncode(v int64) uint64 {
	return uint64(v<<1) ^ uint64(v>>63)
}

func zigzagDecode(v uint64) int64 {
	return int64(v>>1) ^ -int64(v&1)
}

func writeValue(buf *bytes.Buffer, val Value, dict map[uint32]string) error {
	buf.WriteByte(val.Type)
	switch val.Type {
	case TypeNull:
		// nothing
	case TypeBool:
		if val.BoolVal {
			buf.WriteByte(1)
		} else {
			buf.WriteByte(0)
		}
	case TypeInt:
		writeVarint(buf, zigzagEncode(val.IntVal))
	case TypeFloat:
		var fBytes [8]byte
		binary.BigEndian.PutUint64(fBytes[:], math.Float64bits(val.FloatVal))
		buf.Write(fBytes[:])
	case TypeString:
		// Find ID in dict
		id, ok := findKeyByValue(dict, val.StringVal)
		if !ok {
			return errors.New("string não encontrada no dicionário")
		}
		writeVarint(buf, uint64(id))
	case TypeBytes:
		writeVarint(buf, uint64(len(val.BytesVal)))
		buf.Write(val.BytesVal)
	case TypeArray:
		writeVarint(buf, uint64(len(val.ArrayVal)))
		for _, item := range val.ArrayVal {
			if err := writeValue(buf, item, dict); err != nil {
				return err
			}
		}
	case TypeMap:
		writeVarint(buf, uint64(len(val.MapVal)))
		for k, v := range val.MapVal {
			id, ok := findKeyByValue(dict, k)
			if !ok {
				return errors.New("chave de map não encontrada no dicionário")
			}
			writeVarint(buf, uint64(id))
			if err := writeValue(buf, v, dict); err != nil {
				return err
			}
		}
	case TypeDate:
		writeVarint(buf, zigzagEncode(val.DateVal))
	case TypeUUID:
		buf.Write(val.UUIDVal[:])
	default:
		return errors.New("tipo desconhecido")
	}
	return nil
}

func readValue(data []byte, offset *int, dict map[uint32]string) (Value, error) {
	var val Value
	if *offset >= len(data) {
		return val, io.ErrUnexpectedEOF
	}
	val.Type = data[*offset]
	*offset++
	switch val.Type {
	case TypeNull:
	case TypeBool:
		val.BoolVal = data[*offset] != 0
		*offset++
	case TypeInt:
		encoded, err := readVarint(data, offset)
		if err != nil {
			return val, err
		}
		val.IntVal = zigzagDecode(encoded)
	case TypeFloat:
		if *offset+8 > len(data) {
			return val, io.ErrUnexpectedEOF
		}
		bits := binary.BigEndian.Uint64(data[*offset : *offset+8])
		val.FloatVal = math.Float64frombits(bits)
		*offset += 8
	case TypeString:
		id, err := readVarint(data, offset)
		if err != nil {
			return val, err
		}
		s, ok := dict[uint32(id)]
		if !ok {
			return val, errors.New("string não encontrada no dicionário")
		}
		val.StringVal = s
	case TypeBytes:
		length, err := readVarint(data, offset)
		if err != nil {
			return val, err
		}
		if *offset+int(length) > len(data) {
			return val, io.ErrUnexpectedEOF
		}
		val.BytesVal = data[*offset : *offset+int(length)]
		*offset += int(length)
	case TypeArray:
		count, err := readVarint(data, offset)
		if err != nil {
			return val, err
		}
		val.ArrayVal = make([]Value, count)
		for i := uint64(0); i < count; i++ {
			item, err := readValue(data, offset, dict)
			if err != nil {
				return val, err
			}
			val.ArrayVal[i] = item
		}
	case TypeMap:
		count, err := readVarint(data, offset)
		if err != nil {
			return val, err
		}
		val.MapVal = make(map[string]Value)
		for i := uint64(0); i < count; i++ {
			keyID, err := readVarint(data, offset)
			if err != nil {
				return val, err
			}
			key, ok := dict[uint32(keyID)]
			if !ok {
				return val, errors.New("chave de map não encontrada")
			}
			value, err := readValue(data, offset, dict)
			if err != nil {
				return val, err
			}
			val.MapVal[key] = value
		}
	case TypeDate:
		encoded, err := readVarint(data, offset)
		if err != nil {
			return val, err
		}
		val.DateVal = zigzagDecode(encoded)
	case TypeUUID:
		if *offset+16 > len(data) {
			return val, io.ErrUnexpectedEOF
		}
		copy(val.UUIDVal[:], data[*offset:*offset+16])
		*offset += 16
	default:
		return val, errors.New("tipo desconhecido")
	}
	return val, nil
}

func findKeyByValue(m map[uint32]string, target string) (uint32, bool) {
	for id, s := range m {
		if s == target {
			return id, true
		}
	}
	return 0, false
}

File: sdks/rust/src/rodo.rs
use std::collections::HashMap;
use std::fs::{File, OpenOptions};
use std::io::{Read, Seek, SeekFrom, Write};
use std::path::Path;

pub const MAGIC: u32 = 0x52443031;
const VERSION_MAJOR: u8 = 1;
const VERSION_MINOR: u8 = 0;

#[derive(Debug, Clone, PartialEq)]
pub enum Value {
    Null,
    Bool(bool),
    Int(i64),
    Float(f64),
    String(String),
    Bytes(Vec<u8>),
    Array(Vec<Value>),
    Map(HashMap<String, Value>),
    Date(i64),
    Uuid([u8; 16]),
}

impl Value {
    fn type_code(&self) -> u8 {
        match self {
            Value::Null => 0x00,
            Value::Bool(_) => 0x01,
            Value::Int(_) => 0x02,
            Value::Float(_) => 0x03,
            Value::String(_) => 0x04,
            Value::Bytes(_) => 0x05,
            Value::Array(_) => 0x06,
            Value::Map(_) => 0x07,
            Value::Date(_) => 0x08,
            Value::Uuid(_) => 0x09,
        }
    }
}

pub struct Rodo {
    file: File,
    symbols: HashMap<String, u32>,
    symbol_id_counter: u32,
    dict: HashMap<u32, String>,
    dict_id_counter: u32,
    // Single block for simplicity (block_id = 1)
    pairs: HashMap<u32, Value>,
    dirty: bool,
}

impl Rodo {
    pub fn link<P: AsRef<Path>>(path: P) -> std::io::Result<Self> {
        let path = path.as_ref();
        let file = OpenOptions::new()
            .read(true)
            .write(true)
            .create(true)
            .open(path)?;

        let mut rodo = Rodo {
            file,
            symbols: HashMap::new(),
            symbol_id_counter: 0,
            dict: HashMap::new(),
            dict_id_counter: 0,
            pairs: HashMap::new(),
            dirty: false,
        };

        // Check if file is non-empty and load existing data
        let metadata = rodo.file.metadata()?;
        if metadata.len() > 0 {
            rodo.load()?;
        }

        Ok(rodo)
    }

    fn load(&mut self) -> std::io::Result<()> {
        // Simplified: read all data into memory and parse
        let mut data = Vec::new();
        self.file.seek(SeekFrom::Start(0))?;
        self.file.read_to_end(&mut data)?;

        // Parse header
        if data.len() < 18 {
            return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "Arquivo .rd inválido"));
        }
        let magic = u32::from_be_bytes(data[0..4].try_into().unwrap());
        if magic != MAGIC {
            return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "Magic number inválido"));
        }
        // Skip header details for now
        let mut offset = 18;

        // Read symbol table
        let count = read_varint(&data, &mut offset)? as usize;
        for _ in 0..count {
            let _id = read_varint(&data, &mut offset)? as u32;
            let name_len = read_varint(&data, &mut offset)? as usize;
            let name = String::from_utf8(data[offset..offset+name_len].to_vec())
                .map_err(|_| std::io::Error::new(std::io::ErrorKind::InvalidData, "Nome de símbolo inválido"))?;
            offset += name_len;
            let type_hint = data[offset]; offset += 1;
            let _flags = data[offset]; offset += 1;

            self.symbols.insert(name.clone(), self.symbol_id_counter + 1);
            self.symbol_id_counter += 1;
        }

        // Read dictionary
        let count = read_varint(&data, &mut offset)? as usize;
        for _ in 0..count {
            let _id = read_varint(&data, &mut offset)? as u32;
            let str_len = read_varint(&data, &mut offset)? as usize;
            let s = String::from_utf8(data[offset..offset+str_len].to_vec())
                .map_err(|_| std::io::Error::new(std::io::ErrorKind::InvalidData, "String do dicionário inválida"))?;
            offset += str_len;
            self.dict.insert(self.dict_id_counter + 1, s);
            self.dict_id_counter += 1;
        }

        // Read blocks (we assume single block)
        let block_count = read_varint(&data, &mut offset)? as usize;
        if block_count > 0 {
            let _block_id = read_varint(&data, &mut offset)? as u32;
            let pair_count = read_varint(&data, &mut offset)? as usize;
            for _ in 0..pair_count {
                let symbol_id = read_varint(&data, &mut offset)? as u32;
                let value = read_value(&data, &mut offset, &self.dict)?;
                self.pairs.insert(symbol_id, value);
            }
        }

        Ok(())
    }

    pub fn set(&mut self, key: &str, value: Value) -> std::io::Result<()> {
        let symbol_id = match self.symbols.get(key) {
            Some(&id) => id,
            None => {
                self.symbol_id_counter += 1;
                self.symbols.insert(key.to_string(), self.symbol_id_counter);
                self.symbol_id_counter
            }
        };
        self.pairs.insert(symbol_id, value);
        self.dirty = true;
        Ok(())
    }

    pub fn get(&self, key: &str) -> Option<&Value> {
        self.symbols.get(key).and_then(|&id| self.pairs.get(&id))
    }

    pub fn has(&self, key: &str) -> bool {
        self.symbols.contains_key(key) && self.pairs.contains_key(&self.symbols[key])
    }

    pub fn delete(&mut self, key: &str) -> std::io::Result<bool> {
        if let Some(&id) = self.symbols.get(key) {
            if self.pairs.remove(&id).is_some() {
                self.dirty = true;
                return Ok(true);
            }
        }
        Ok(false)
    }

    pub fn keys(&self) -> Vec<String> {
        self.pairs.keys()
            .filter_map(|&id| self.symbols.iter().find(|(_, &sid)| sid == id).map(|(name, _)| name.clone()))
            .collect()
    }

    pub fn all(&self) -> HashMap<String, &Value> {
        let mut result = HashMap::new();
        for (name, &id) in &self.symbols {
            if let Some(val) = self.pairs.get(&id) {
                result.insert(name.clone(), val);
            }
        }
        result
    }

    pub fn close(mut self) -> std::io::Result<()> {
        if self.dirty {
            self.save()?;
        }
        // File is closed on drop
        Ok(())
    }

    fn save(&mut self) -> std::io::Result<()> {
        let mut data = Vec::new();

        // Header
        data.extend_from_slice(&MAGIC.to_be_bytes());
        data.push(VERSION_MAJOR);
        data.push(VERSION_MINOR);
        data.push(0); // flags
        // Timestamp (8 bytes) - use current time as simple value
        let timestamp = 0u64;
        data.extend_from_slice(&timestamp.to_be_bytes());
        // Header CRC32 (simplified: 0)
        data.extend_from_slice(&[0u8; 4]);

        // Symbol table
        write_varint(&mut data, self.symbols.len() as u64)?;
        for (name, &id) in &self.symbols {
            write_varint(&mut data, id as u64)?;
            let name_bytes = name.as_bytes();
            write_varint(&mut data, name_bytes.len() as u64)?;
            data.extend_from_slice(name_bytes);
            // type_hint: we don't store type hints, use 0
            data.push(0);
            data.push(0); // flags
        }

        // Dictionary: collect strings from values that are strings or map keys
        let mut dict_map: HashMap<String, u32> = HashMap::new();
        for value in self.pairs.values() {
            collect_strings(value, &mut dict_map, &mut self.dict_id_counter)?;
        }
        // Override dict with collected
        self.dict.clear();
        for (s, id) in &dict_map {
            self.dict.insert(*id, s.clone());
        }
        write_varint(&mut data, self.dict.len() as u64)?;
        let mut dict_entries: Vec<_> = self.dict.iter().collect();
        dict_entries.sort_by_key(|(id, _)| **id);
        for (id, s) in dict_entries {
            write_varint(&mut data, *id as u64)?;
            let bytes = s.as_bytes();
            write_varint(&mut data, bytes.len() as u64)?;
            data.extend_from_slice(bytes);
        }

        // Block (single block)
        write_varint(&mut data, 1)?; // block_count
        write_varint(&mut data, 1)?; // block_id
        write_varint(&mut data, self.pairs.len() as u64)?;
        for (symbol_id, value) in &self.pairs {
            write_varint(&mut data, *symbol_id as u64)?;
            write_value(&mut data, value, &dict_map)?;
        }

        // Metadata (simplified)
        write_varint(&mut data, self.pairs.len() as u64)?;
        write_varint(&mut data, self.symbols.len() as u64)?;
        write_varint(&mut data, self.dict.len() as u64)?;
        write_varint(&mut data, data.len() as u64)?;
        // CRC32 (0)
        data.extend_from_slice(&[0u8; 4]);

        self.file.seek(SeekFrom::Start(0))?;
        self.file.write_all(&data)?;
        self.file.set_len(data.len() as u64)?;
        self.file.flush()?;
        self.dirty = false;
        Ok(())
    }
}

fn collect_strings(value: &Value, dict: &mut HashMap<String, u32>, counter: &mut u32) -> std::io::Result<()> {
    match value {
        Value::String(s) => {
            if !dict.contains_key(s) {
                *counter += 1;
                dict.insert(s.clone(), *counter);
            }
        }
        Value::Map(map) => {
            for (k, v) in map {
                if !dict.contains_key(k) {
                    *counter += 1;
                    dict.insert(k.clone(), *counter);
                }
                collect_strings(v, dict, counter)?;
            }
        }
        Value::Array(arr) => {
            for v in arr {
                collect_strings(v, dict, counter)?;
            }
        }
        _ => {}
    }
    Ok(())
}

fn read_varint(data: &[u8], offset: &mut usize) -> std::io::Result<u64> {
    let mut result = 0u64;
    let mut shift = 0;
    loop {
        if *offset >= data.len() {
            return Err(std::io::Error::new(std::io::ErrorKind::UnexpectedEof, "Varint incompleto"));
        }
        let byte = data[*offset];
        *offset += 1;
        result |= ((byte & 0x7F) as u64) << shift;
        shift += 7;
        if byte & 0x80 == 0 {
            break;
        }
        if shift > 70 {
            return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "Varint muito longo"));
        }
    }
    Ok(result)
}

fn write_varint(out: &mut Vec<u8>, value: u64) -> std::io::Result<()> {
    let mut v = value;
    while v >= 0x80 {
        out.push(((v & 0x7F) as u8) | 0x80);
        v >>= 7;
    }
    out.push(v as u8);
    Ok(())
}

fn zigzag_encode(v: i64) -> u64 {
    ((v << 1) ^ (v >> 63)) as u64
}

fn zigzag_decode(v: u64) -> i64 {
    ((v >> 1) as i64) ^ -((v & 1) as i64)
}

fn read_value(data: &[u8], offset: &mut usize, dict: &HashMap<u32, String>) -> std::io::Result<Value> {
    let type_code = data[*offset];
    *offset += 1;
    match type_code {
        0x00 => Ok(Value::Null),
        0x01 => {
            let b = data[*offset] != 0;
            *offset += 1;
            Ok(Value::Bool(b))
        }
        0x02 => {
            let encoded = read_varint(data, offset)?;
            Ok(Value::Int(zigzag_decode(encoded)))
        }
        0x03 => {
            let bytes: [u8; 8] = data[*offset..*offset+8].try_into().unwrap();
            *offset += 8;
            Ok(Value::Float(f64::from_be_bytes(bytes)))
        }
        0x04 => {
            let id = read_varint(data, offset)? as u32;
            let s = dict.get(&id).cloned().unwrap_or_default();
            Ok(Value::String(s))
        }
        0x05 => {
            let len = read_varint(data, offset)? as usize;
            let bytes = data[*offset..*offset+len].to_vec();
            *offset += len;
            Ok(Value::Bytes(bytes))
        }
        0x06 => {
            let count = read_varint(data, offset)? as usize;
            let mut arr = Vec::with_capacity(count);
            for _ in 0..count {
                arr.push(read_value(data, offset, dict)?);
            }
            Ok(Value::Array(arr))
        }
        0x07 => {
            let count = read_varint(data, offset)? as usize;
            let mut map = HashMap::new();
            for _ in 0..count {
                let key_id = read_varint(data, offset)? as u32;
                let key = dict.get(&key_id).cloned().unwrap_or_default();
                let val = read_value(data, offset, dict)?;
                map.insert(key, val);
            }
            Ok(Value::Map(map))
        }
        0x08 => {
            let encoded = read_varint(data, offset)?;
            Ok(Value::Date(zigzag_decode(encoded)))
        }
        0x09 => {
            let mut uuid = [0u8; 16];
            uuid.copy_from_slice(&data[*offset..*offset+16]);
            *offset += 16;
            Ok(Value::Uuid(uuid))
        }
        _ => Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "Tipo desconhecido")),
    }
}

fn write_value(out: &mut Vec<u8>, value: &Value, dict: &HashMap<String, u32>) -> std::io::Result<()> {
    out.push(value.type_code());
    match value {
        Value::Null => {}
        Value::Bool(b) => out.push(if *b { 1 } else { 0 }),
        Value::Int(i) => write_varint(out, zigzag_encode(*i))?,
        Value::Float(f) => out.extend_from_slice(&f.to_be_bytes()),
        Value::String(s) => {
            let id = dict.get(s).copied().unwrap_or(0);
            write_varint(out, id as u64)?;
        }
        Value::Bytes(b) => {
            write_varint(out, b.len() as u64)?;
            out.extend_from_slice(b);
        }
        Value::Array(arr) => {
            write_varint(out, arr.len() as u64)?;
            for v in arr {
                write_value(out, v, dict)?;
            }
        }
        Value::Map(map) => {
            write_varint(out, map.len() as u64)?;
            for (k, v) in map {
                let id = dict.get(k).copied().unwrap_or(0);
                write_varint(out, id as u64)?;
                write_value(out, v, dict)?;
            }
        }
        Value::Date(d) => write_varint(out, zigzag_encode(*d))?,
        Value::Uuid(u) => out.extend_from_slice(u),
    }
    Ok(())
}

pub fn link<P: AsRef<Path>>(path: P) -> std::io::Result<Rodo> {
    Rodo::link(path)
}